### PR TITLE
Histogram: correction when dealing with log-scale with negative values

### LINKED ISSUE #815: An error when plotting histogram
Reported bug @eivindsm 
```script
An error occurred during plotting. This stack trace is helpful for diagnosing the problem.--------------------------------------------------------------------------------
  File "/prog/res/komodo/2020.04.rc2-py36/root/lib/python3.6/site-packages/ert_gui/tools/plot/plot_widget.py", line 73, in updatePlot
    self._plotter.plot(self._figure, plot_context, case_to_data_map, observations)
  File "/prog/res/komodo/2020.04.rc2-py36/root/lib/python3.6/site-packages/ert_gui/plottery/plots/histogram.py", line 14, in plot
    plotHistogram(figure, plot_context, case_to_data_map, _observation_data)
  File "/prog/res/komodo/2020.04.rc2-py36/root/lib/python3.6/site-packages/ert_gui/plottery/plots/histogram.py", line 90, in plotHistogram
    _plotHistogram(axes[case], config, data[case], case, bin_count, use_log_scale, minimum, maximum)
  File "/prog/res/komodo/2020.04.rc2-py36/root/lib/python3.6/site-packages/ert_gui/plottery/plots/histogram.py", line 153, in _plotHistogram
    bins = _histogramLogBins(bin_count, minimum, maximum)
  File "/prog/res/komodo/2020.04.rc2-py36/root/lib/python3.6/site-packages/ert_gui/plottery/plots/histogram.py", line 177, in _histogramLogBins
    minimum = log10(float(minimum))
Exception type: ValueError
math domain error
```
We should double-check that the minimum's computed correctly:
https://github.com/equinor/ert/blob/a035f988a5a85ddc6a5ee108a9205c356a31f404/ert_gui/plottery/plots/histogram.py#L61

### PR DESCRIPTION
**Pull request by xjules**
**Issue**
Resolves #815 

**Approach**
Different way to compute min and max. Do not set limits when minimum is negative.

### CODE REVIEW

File: ert_gui/plottery/plots/histogram.py [hunk 1]
```diff
@@ -178,8 +179,8 @@ def _histogramLogBins(bin_count, minimum=None, maximum=None):
     @type data: pandas.DataFrame
     @rtype: int
     """
-    minimum = log10(float(minimum))
-    maximum = log10(float(maximum))
+    minimum = numpy.sign(float(minimum)) * log10(fabs(minimum))
```
Comment (lars-petter-hauge) [inline ert_gui/plottery/plots/histogram.py:182]: I think you mentioned the case of minimum being 0 when we talked, but how does this help against that? `fabs` of 0 is still 0 :)

File: ert_gui/plottery/plots/histogram.py [hunk 2]
```diff
@@ -166,7 +166,8 @@ def _plotHistogram(axes, plot_config, data, label, bin_count, use_log_scale=Fals
         minimum -= 0.5
         maximum += 0.5
 
-    axes.set_xlim(minimum, maximum)
+    if not use_log_scale or minimum>0:
```
Comment (lars-petter-hauge) [inline ert_gui/plottery/plots/histogram.py:169]: The condition here should really never be `False`, we should always enter the `if` statement. I'm not completely sure, but if setting limits to auto is the same as not setting them in the first place, then I would prefer:

```
axes.set_xlim(minimum, maximum)

if use_log_scale and minimum <= 0:
    axes.set_xlim(auto=True)
```
As that is the special case.

But is it necessary at all? From looking at the changes in `_histogramLogBins`, it seems we're trying to incorporate the negative value, and hence setting the limit non the less makes sense

Comment (xjules) [reply ert_gui/plottery/plots/histogram.py:169]: Setting the negative limit(s) for `log_scale` axis will through an error or a warning at least, therefore I would then suggest:
```python
if use_log_scale and minimum <= 0:
    axes.set_xlim(auto=True)
else:
    axes.set_xlim(minimum, maximum)
```

File: ert_gui/plottery/plots/histogram.py (hunk 1 above)
Comment (xjules) [reply ert_gui/plottery/plots/histogram.py:182]: Well spotted. Do you think we should do something like this then?:
```python
if np.isclose(mimimum, 0):
   minimum = minumum + 0.0000001
```

File: ert_gui/plottery/plots/histogram.py (hunk 2 above)
Comment (lars-petter-hauge) [reply ert_gui/plottery/plots/histogram.py:169]: Ah, of course, there's a `log_scale` axis, then I completely agree. Agree with the proposition

File: ert_gui/plottery/plots/histogram.py (hunk 1 above)
Comment (lars-petter-hauge) [reply ert_gui/plottery/plots/histogram.py:182]: Is it possible to get number in the range 10E-12 - 10E-15?

In ert, if you really need them, you'll have to transform them after getting the range from e.g. 10E0-10E-3?

**Comment by lars-petter-hauge** [comment]
My concern is that we kind of replace a bug (exception raised) with an ignore. Although it runs, it just hides the problem.

With this fix, all numbers lower than 1 will be missing from the plot - This should be solved differently.

**Comment by xjules** [comment]
> My concern is that we kind of replace a bug (exception raised) with an ignore. Although it runs, it just hides the problem.
> 
> With this fix, all numbers lower than 1 will be missing from the plot - This should be solved differently.

That's indeed true, but then we would need to change log axis as this one doesn't accept negative values.

**Comment by lars-petter-hauge** [comment]
The configuration of `LOGNORMAL` distribution describes the logarithmic of the value. If we have a distribution as such:

`KEY_NAME LOGNORMAL -5 1`

We will get values around 0.00001 (`log10(-5)`). The values in the GUI when selecting a parameter prefixed with `LOG10_`, will get you the logarithm of the distributed values, so in our case negative values. These should not be plotted on a logarithmic scale. Instead the values without the prefix `LOG10_` are the ones that should be plotted on a logarithmic scale, and the values _with_the prefix should be plotted on a normal scale.

**Comment by xjules** [comment]
> The configuration of `LOGNORMAL` distribution describes the logarithmic of the value. If we have a distribution as such:
> 
> `KEY_NAME LOGNORMAL -5 1`
> 
> We will get values around 0.00001 (`log10(-5)`). 

When using `KEY_NAME LOGNORMAL -5 1` what we get actually is this:
```python
x = np.random.rand(10)
# Our values in distribution based on https://github.com/equinor/libres/blob/3585997ac0a51d10008307f16c86182b93433af2/lib/enkf/trans_func.cpp#L153:
data = np.exp(x * 1 - 5)
data
Out: array([0.01062894, 0.00687533, 0.01588024, 0.01035408, 0.00949111,
       0.01101723, 0.01314581, 0.01080298, 0.00806015, 0.00933471])
```

> The values in the GUI when selecting a parameter prefixed with `LOG10_`, will get you the logarithm of the distributed values, so in our case negative values. These should not be plotted on a logarithmic scale. Instead the values without the prefix `LOG10_` are the ones that should be plotted on a logarithmic scale, and the values _with_the prefix should be plotted on a normal scale.

Showing `LOG10_` is correct, then only part what's missing are those values which are `<1` since log_scale in matplotlib does not support those. So nothing is swapped in my understanding at least.

> 
> If there are any metadata related to the values occurring from a logarithmic distribution, that would be great. Would rather avoid having to search if there are any keys with the same name that has a prefix of `LOG10_` to determine the axis scale type.

Agree with this.

**Comment by lars-petter-hauge** [comment]
I still think that the axis are swapped, i.e. the parameter with the prefix `LOG_` should _not_ be plotted on a logarithmic scale, while any parameter that comes from a logarithmic distribution should (unless prefixed with `LOG_` be plotted on a logarithmic scale.

I checked out master (not including this PR) and swapped them around.

Then I made an example with two keys: `POS_LOG` and `NEG_LOG`, with the following distributions:
`POS_LOG LOGNORMAL 5 2`
`NEG_LOG LOGNORMAL -5 2`

I would in this case expect the values for
- `POS_LOG` to be within 10^3 -> 10^7, mean at 10^5.
- `NEG_LOG` to be within 10^-7 -> 10^-3, mean at 10^-5.

Now plotting:

POS_LOG
LOG10_POS_LOG

I think that looks pretty good.

For the low values:

NEG_LOG
LOG10_NEG_LOG

-------------------------

The expectation weren't really correct, they more look like the first number describes the min/max value, and it spans around 1. The std looks correct.

Regardless, I think the plots look as expected to be honest.

**Comment by lars-petter-hauge** [comment]
> If there are any metadata related to the values occurring from a logarithmic distribution, that would be great. Would rather avoid having to search if there are any keys with the same name that has a prefix of LOG10_ to determine the axis scale type.

To partly answer my own question, I couldn't find any metadata as per now. But this is how I think this should be solved:

The `PlotContext` seems to be the most appropriate candidate to know more about the data. I would add the metadata whether it should be plotted on a log scale or not here.

https://github.com/equinor/ert/blob/master/ert_gui/plottery/plot_context.py

This plot context is initialized in the `PlotWindow`
https://github.com/equinor/ert/blob/b369817db41e9918457c55ff39b2a2c866a04d99/ert_gui/tools/plot/plot_window.py#L103

Now the `key_def`:
https://github.com/equinor/ert/blob/b369817db41e9918457c55ff39b2a2c866a04d99/ert_gui/tools/plot/plot_window.py#L85

is populated in the `plot_api`:
https://github.com/equinor/ert/blob/b369817db41e9918457c55ff39b2a2c866a04d99/ert_gui/tools/plot/plot_api.py#L13

Here is where I would add another entry to the key_def. If this really has to be made as checking whether there exists another key with `LOG_` prefix, then be it. For the new storage that is coming, we could pick the most appropriate solution (perhaps we could get the distribution for a parameter and do the checking that way).

**Comment by xjules** [comment]
> I still think that the axis are swapped, i.e. the parameter with the prefix `LOG_` should _not_ be plotted on a logarithmic scale, while any parameter that comes from a logarithmic distribution should (unless prefixed with `LOG_` be plotted on a logarithmic scale.

So I did some simple experiments on my own and indeed they are swapped. Simple example:
```python
import matplotlib.pyplot as plt
import numpy as np

data = 10000 * np.random.rand(100)
plt.subplot(121)
plt.xscale('log')
# Left: bins are like 10**(np.arange(0,5))
plt.hist(data)

plt.subplot(122)
# Right: bins are like 20000 * np.arange(0,5)
plt.hist(data)
plt.show()
```
Figure_1

**Comment by lars-petter-hauge** [comment]
Having a checkbox to select scale is a good idea. In most cases though I think that the log scale will be used only for the logarithmic values, and in that sense it would involve an unnecessary click (and another for unchecking the box when switching to next). It's pro's and con's.. 

But this is a feature request more than a bug fix. I suggest that we fix the bug first before anything else.

> The solution when searching for another key with LOG10_ prefix is doable but not as nice, especially since we populate new keys with LOG10_ just for the sake of visualization

The `LOG10_` prefix comes from libres. You can see in the parameters.json that they are already included. The ERT GUI only populates the widgets based on the keys, it doesn't, from my recollection, do anything in advance with them.

> The solution when searching for another key with LOG10_ prefix is doable but not as nice

I agree. But I'm pretty sure that we will have the distribution information added to the parameters when asking for them in the new storage solution. In that case the `StorageClient` API, that exposes the same functions as the `PlotAPI`:
https://github.com/equinor/ert/blob/8e2efee9991dfd5679fa2c8aa4afb41a0d7ade62/ert_shared/storage/client.py#L37
will provide the details. I'm not sure about the entrypoint just yet, but we'll figure it out.

So this would only be the solution for the old storage, and it will be added to the section that will be deprecated.

**Pull request merged**